docs/source/user/jitclass.rst

```diff
@@ -40,11 +40,49 @@ initializing each defined fields.  Uninitialized fields contains garbage data.
 Methods and properties (getters and setters only) can be defined.  They will be
 automatically compiled.
 
-Specifying ``numba.typed`` containers as class members
-======================================================
-It is often desirable to use a ``numba.typed.Dict`` or a ``numba.typed.List`` as
-a class member in a ``jitclass``. Methods for using these types and various
-common patterns are presented in the following:
+
+Inferred class member types with ``as_numba_type``
+==================================================
+
+Fields of a ``jitclass`` can also be inferred from Python type annotations.
+
+.. literalinclude:: ../../../numba/tests/doc_examples/test_jitclass.py
+   :language: python
+   :start-after: magictoken.ex_jitclass_type_hints.begin
+   :end-before: magictoken.ex_jitclass_type_hints.end
+   :dedent: 8
+
+Any type annotations on the class will be used to extend the spec if that field
+is not already present.  The Numba type corresponding to given Python type is
+inferred using ``as_numba_type``.  For example, if we have the class
+
+.. code-block:: python
+
+    @jitclass([("x", int32)])
+    class Foo:
+        x: int
+        y: float
+        z: SomeOtherType
+
+        def __init__(self, x: int, y: float, z: SomeOtherType):
+            ...
+
+then the full spec used for ``Foo`` will be:
+
+* ``"x": int32`` (specified in the original ``spec`` passed to ``jitclass``)
```

Comment:
> it's possible to make having type annotations bans the use of the spec

I'm not sure if this would work for numpy types.  Any array would get a type annotation `np.ndarray`, but numba requires more information.  In a future where numpy has generic type annotations for array this could be solved, but I don't think we're there yet.  This situation is why I thought to allow overwriting annotations with the `spec`.

If you remove the `@jitclass` decorator from the above example, that's still not valid Python code (also the type annotations for `Counter` seem incorrect).  What are you suggesting to change because of this example?  Add a check that every field is initialized in `__init__`?

Reply:
You convinced me. We need to have both spec and annotations. 

> Add a check that every field is initialized in __init__?

We'll need to address the segfault in a different PR. Prevent the segfault or check that all fields are initialized, or both.

Reply:
Ok so is there any action item for this comment thread (beyond the documentation updates)?